feather/plugin-host/src/command_dispatcher.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    cell::RefCell,
+    convert::TryInto,
+    fmt::{self, Debug},
+    rc::Rc,
+};
+
+use crate::{context::PluginPtrMut, Game, PluginId, PluginManager};
+use anyhow::bail;
+use feather_ecs::{Entity, EntityRef};
+use lieutenant::regex::early_termination::CmdPos;
+use lieutenant::regex::{dfa::DFA, NFA};
+use quill_common::EntityId;
+
+// Constant that should reflect how many commands a default
+// server starts with. Vanilla has roughly 60
+const NUM_COMMANDS: usize = 20;
+
+type CommandId = u32;
+
+/// A collection of data containing what plugin a command was defined in,
+/// the pointer to its Box<Command<...>>, and the regex that determines
+/// when the command should be called.(The command does some parsing beyond the regex).
+pub struct CommandDescription {
+    pub plugin_id: PluginId,
+    pub regex: String,
+    // Should only come from host_call register_system.
+    // it is a pointer to a Box<dyn Command<GameState=(Plugin,CommandContext),z CommandResult=i64>>
+    pub callback: PluginPtrMut<u8>,
+}
+
+impl CommandDescription {
+    // Returns true or false depending on if the command executed or not.
+    // If it returns false that means the commands parser did not recognize
+    // the commnd.
+    /// caller: None => caller was terminal,
+    /// caller: Some(x) => player x called command.
+    fn call(&self, game: &mut Game, input: &str, caller: Option<EntityId>) -> anyhow::Result<()> {
+        let plugin_manager = Rc::clone(&*game.resources.get::<Rc<RefCell<PluginManager>>>()?);
+        let plugin_manager = plugin_manager.borrow();
+        let plugin = plugin_manager.plugin(self.plugin_id);
+
+        match plugin {
+            Some(plugin) => {
+                plugin.call_command(game, input, self.callback, caller)?;
+                Ok(())
+            }
+            None => {
+                // Plugin has been unloaded.
+                bail!("Plugin was not loaded.")
+            }
+        }
+    }
+}
+
+pub struct CommandDispatcher {
+    dfa: DFA<CmdPos<CommandId>>,
+    commands: Vec<CommandDescription>,
+}
+
+impl fmt::Debug for CommandDispatcher {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.dfa.fmt(f)
+    }
+}
+
+impl CommandDispatcher {
+    pub fn new() -> Self {
+        Self {
+            dfa: DFA::new(),
+            commands: Vec::with_capacity(NUM_COMMANDS),
```

Comment:
I'm again skeptical of having this as a sort of magic constant, then again this is just the initial capacity, so it's probably fine.

Reply:
The default capacity of `Vec` should be fine.